presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java

```diff
@@ -224,50 +281,184 @@ public void addInput(Page page)
+        if (processPendingInput()) {
+            state = State.HAS_OUTPUT;
+        }
         operatorContext.setMemoryReservation(pagesIndex.getEstimatedSize().toBytes());
     }
 
+    /**
+     * @return true if a full group has been buffered after processing the pendingInput, false otherwise
+     */
+    private boolean processPendingInput()
+    {
+        checkState(pendingInput != null);
+        pendingInput = updatePagesIndex(pendingInput);
+
+        if (pendingInput != null || state == State.FINISHING) {
+            // Seeing unused input or the end of input implies that we have currently buffered a full group
+            sortPagesIndexIfNecessary();
+            return true;
+        }
+        else {
+            return false;
+        }
+    }
+
+    /**
+     * @return the unused section of the page, or null if fully applied.
+     * pagesIndex guaranteed to have at least one row after this method returns
+     */
+    private Page updatePagesIndex(Page page)
+    {
+        checkArgument(page.getPositionCount() > 0);
+
+        // TODO: Fix pagesHashStrategy to allow specifying channels for comparison, it currently requires us to rearrange the right side blocks in consecutive channel order
+        Page preGroupedPage = rearrangePage(page, preGroupedChannels);
+        if (pagesIndex.getPositionCount() == 0 || pagesIndex.positionEqualsRow(preGroupedPartitionHashStrategy, 0, 0, preGroupedPage.getBlocks())) {
+            // find the position where the pre-grouped columns change
+            int partitionEnd = findGroupEnd(preGroupedPage, preGroupedPartitionHashStrategy, 0);
+
+            // add the section of the page that contains values for the current partition
+            pagesIndex.addPage(page.getRegion(0, partitionEnd));
+
+            // save the remaining page, which may contain multiple partitions
+            if (page.getPositionCount() - partitionEnd > 0) {
+                return page.getRegion(partitionEnd, page.getPositionCount() - partitionEnd);
+            }
+            else {
+                return null;
+            }
+        }
+        else {
+            // We had previous results buffered, but the new page did not start with the same pre-grouped column values
+            return page;
+        }
+    }
+
+    private static Page rearrangePage(Page page, int[] channels)
+    {
+        Block[] newBlocks = new Block[channels.length];
+        for (int i = 0; i < channels.length; i++) {
+            newBlocks[i] = page.getBlock(channels[i]);
+        }
+        return new Page(page.getPositionCount(), newBlocks);
+    }
+
     @Override
     public Page getOutput()
     {
-        if (state != State.HAS_OUTPUT) {
+        Page page = extractOutput();
+        operatorContext.setMemoryReservation(pagesIndex.getEstimatedSize().toBytes());
+        return page;
+    }
+
+    private Page extractOutput()
+    {
+        if (state == State.NEEDS_INPUT || state == State.FINISHED) {
```

Comment:
you might want to leave this in getOutput so you don't keep updating the memory reservation